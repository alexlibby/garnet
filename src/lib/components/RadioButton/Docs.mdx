import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as RadioButtonStories from "./RadioButton.stories";

<Meta of={RadioButtonStories} />

# RadioButton

RadioButton is the primary component. It has 1 possible states.

- [Default](#default)
- [Disabled](#disabled)

## Default

This is the default state for our RadioButton component:

<Canvas of={RadioButtonStories.Default} />

## Disabled

This is the checkbox component in a disabled state:

<Canvas of={RadioButtonStories.Disabled} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={RadioButtonStories} />
