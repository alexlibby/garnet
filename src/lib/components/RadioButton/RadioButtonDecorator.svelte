<div>
  <slot />
</div>
<div id="message">Selected option: (not selected)</div>

<style>
  div {
    margin-top: 10px;
  }

  #message {
    font-family: Arial, Helvetica, sans-serif;
  }
</style>