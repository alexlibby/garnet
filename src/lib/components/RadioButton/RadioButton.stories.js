import RadioButton from "./RadioButton.svelte";
import RadioButtonDecorator from "./RadioButtonDecorator.svelte";
import { action } from "@storybook/addon-actions";
import { fn } from "@storybook/test";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

let selectOptions = [
  {
    value: "Armstrong-Siddeley",
    label: "Armstrong-Siddeley",
  },
  {
    value: "Jaguar Mark II",
    label: "Jaguar Mark II",
  },
  {
    value: "Ford Zephyr",
    label: "Ford Zephyr",
  },
];

export default {
  title: "Garnet UI Library/Basic Components/RadioButton",
  component: RadioButton,
  decorators: [() => RadioButtonDecorator],
  argTypes: {
    options: selectOptions,
    disabled: { control: "boolean" },
  },
  props: {
    disabled: false,
  },
};

export const Default = (args) => ({
  Component: RadioButton,
  props: {
    ...args,
    selectOptions,
  },
  on: {
    change: (event) => {
      const selectedItem = event.target.value;

      document.getElementById("message").innerHTML =
        `Selected option: ${selectedItem}`;
    },
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const Disabled = (args) => ({
  Component: RadioButton,
  props: {
    ...args,
    selectOptions,
    disabled: true,
  },
  on: { change: action("on-change") },
});

Disabled.parameters = {
  badges: [BADGES.ALPHA],
};
