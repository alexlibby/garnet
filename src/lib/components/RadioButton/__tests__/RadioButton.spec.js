import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import RadioButton from "../../RadioButton/RadioButton.svelte";

const RadioButtonHTML = `<fieldset id="garnet" class="s-F2UExF4kPp6W"><div class="legend s-F2UExF4kPp6W" id="label-Legend">Legend</div> <input class="sr-only s-F2UExF4kPp6W" type="radio" id="armstrong-siddeley" role="radio" value="Armstrong-Siddeley"> <label for="armstrong-siddeley" class="s-F2UExF4kPp6W">Armstrong-Siddeley </label><input class="sr-only s-F2UExF4kPp6W" type="radio" id="jaguar-mark-ii" role="radio" value="Jaguar Mark II"> <label for="jaguar-mark-ii" class="s-F2UExF4kPp6W">Jaguar Mark II </label><input class="sr-only s-F2UExF4kPp6W" type="radio" id="ford-zephyr" role="radio" value="Ford Zephyr"> <label for="ford-zephyr" class="s-F2UExF4kPp6W">Ford Zephyr </label></fieldset>`;
expect.extend(AxeMatchers);

describe("Tests for RadioButton", () => {
  const props = {
    legend: "RadioButton Group",
  };

  test("should render properly", () => {
    const result = render(RadioButton, { props });
    expect(() => result).not.toThrow();
  });

  test("show a label with correct text", () => {
    const result = render(RadioButton, { props });
    expect(result).not.toBeNull();
  });

  test("should show a checked radio when clicked", async () => {
    const result = render(RadioButton, { props });
    const RadioButtonsGroup = screen.getAllByRole("radio")[1];
    await fireEvent.click(RadioButtonsGroup);
    expect(RadioButtonsGroup.checked).toBe(true);
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new RadioButton({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => RadioButtonHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
