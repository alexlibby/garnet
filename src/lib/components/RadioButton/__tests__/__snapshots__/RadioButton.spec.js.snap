// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for RadioButton > get a snapshot of component 1`] = `"<fieldset id="garnet" class="s-F2UExF4kPp6W"><div class="legend s-F2UExF4kPp6W" id="label-RadioButton Group">RadioButton Group</div> <input class="sr-only s-F2UExF4kPp6W" type="radio" id="test-slot-1" value="Test slot 1"> <label for="test-slot-1" class="s-F2UExF4kPp6W">Test slot 1 </label><input class="sr-only s-F2UExF4kPp6W" type="radio" id="test-slot-2" value="Test slot 2"> <label for="test-slot-2" class="s-F2UExF4kPp6W">Test slot 2 </label><input class="sr-only s-F2UExF4kPp6W" type="radio" id="test-slot-3" value="Test slot 3"> <label for="test-slot-3" class="s-F2UExF4kPp6W">Test slot 3 </label></fieldset><!--<RadioButton>-->"`;
